Feature/ideation functionality pt2

Dan-Y-Ko opened the pull request:
# Description

- Added banner image when project idea list is empty
- Adjusted styles of the pages to fit the new design specs
- Added confirmation modals
- Refactored banner component
- Did a bit of modal refactor (both functionality and styling)
- Implemented auto-save functionality in the form

## Issue link

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration


# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

JaneMoroz commented on src/app/(main)/my-voyage/[teamId]/ideation/components/IdeationForm.tsx:
```diff
@@ -168,39 +171,110 @@ export default function IdeationForm() {
-  function renderDeleteButtonContent() {
-    if (deleteIdeationLoading) {
+  useEffect(() => {
+    async function autoSave() {
```
I guess we're gonna have this `autoSave` functionality in many places. Is it possible to move it to utils so we can just import and call it from other places (passing some arguments of course)?

JaneMoroz replied:
So basically everytime a user types smth, it gets saved to the database (every 5 seconds)... isn't it too many API calls?

Dan-Y-Ko replied:
Yes ideally we want this as a re-usable function. The reason I didn't do that in this PR is because there's quite a lot going on in this logic and Idk what will be needed / not needed for the other features so I couldn't really come up with a good refactor for this right now. I think we can revisit this in the future when we have a clearer idea of what's common between the different forms.

As for saving to the backend. Yes, that is an issue. However, we don't have a better solution right now unfortunately. As soon as client-side route interceptors get supported, we should migrate to that solution instead of this.

Dan-Y-Ko commented on src/app/(main)/my-voyage/[teamId]/ideation/components/ContributionCard.tsx:
```diff
@@ -35,10 +35,14 @@ function ContributionCard({
+        <Badge
+          title={member.firstName}
+          avatarUrlImage={member.avatar}
+          className="w-full"
```
Yup, if we need to do that we can take care of that. We can just display the first name. That is what I did in my implementation.

Dan-Y-Ko commented on src/store/features/modal/modalSlice.ts:
```diff
@@ -4,46 +4,50 @@ export type ModalType =
   type: ModalType | undefined;
   isOpen: boolean;
   isEditing?: boolean;
+  content: ContentPayload;
```
Yea I was having issues trying to get it to work when I had different slices for the modal and error modal but I realized that isn't really needed and it seems to work now

Dan-Y-Ko commented on src/components/modals/Modal.tsx:
```diff
@@ -89,17 +93,22 @@ export default function Modal({
-            <div className="flex items-center justify-between pb-8">
-              <h3 className="text-xl font-semibold capitalize">{title}</h3>
+            <div
+              className={`flex items-center justify-between px-10 py-6 ${headerBackground}`}
```
https://imgur.com/a/zF786OO

tbh, I don't see a difference 😅 I copied the code you suggested, and it looks the same to me.

I am pushing it up anyways for now

JaneMoroz replied:
The difference is visible in delete confirmation modal 🤓

Dan-Y-Ko commented on src/components/modals/ErrorModal.tsx:
```diff
@@ -1,29 +1,43 @@
+      icon={<ExclamationTriangleIcon className="w-8" />}
+      headerBackground="bg-error-content"
+    >
       <div className="flex flex-col overflow-hidden">
```
We actually don't have any long error messages. The backend has put very generic error messages in the backend and not the exact message.

Dan-Y-Ko replied:
I will push this up anyways though.

JaneMoroz commented:
@Dan-Y-Ko I fixed padding (left/right padding was missing) and removed unnecessary `mt` in `IdeationForm`. Also replaced images with high quality versions.

JaneMoroz commented:
@Dan-Y-Ko There's one error I get in the dev tools console when I add vote or remove vote (but everything seems to work just fine):
```
Cannot update a component (`VoteCard`) while rendering a different component (`IdeationProvider`). To locate the bad setState() call inside `IdeationProvider`...
```
Have you seen it?

Dan-Y-Ko commented:
> @Dan-Y-Ko There's one error I get in the dev tools console when I add vote or remove vote (but everything seems to work just fine):
> 
> ```
> Cannot update a component (`VoteCard`) while rendering a different component (`IdeationProvider`). To locate the bad setState() call inside `IdeationProvider`...
> ```
> 
> Have you seen it?

Yes, it works fine in dev but there was an issue in prod. I fixed it so the warning goes away. There was also a warning for banner image which I also fixed.

JaneMoroz reviewed (approved):
Looks good 👍
